### ISSUE TITLE
PSP relocation types not supported

### ISSUE THREAD

**Issue by John-K** [kotcrab/ghidra-allegrex#7]
It would appear that the custom PSP relocation types aren't supported:
```
#define R_MIPS_X_HI16   13
#define R_MIPS_X_J26    14
#define R_MIPS_X_JAL26  15
```
See https://github.com/pspdev/prxtool/blob/08849d87fa8692f2ba29238ec1710f2706e3e021/elftypes.h#L163-L165 and https://github.com/pspdev/prxtool/blob/08849d87fa8692f2ba29238ec1710f2706e3e021/ProcessPrx.C#L1558-L1635

**Comment by kotcrab** [comment]
Interesting, they are not supported by PPSSPP either so they can't be used in games. I don't think I'll try to implement them without an executable to test it on.

**Comment by John-K** [comment]
It would make sense that they're not implemented in PPSSPP as they're used in System PRXs. A small snippet from `prxtool` output for `lowio.prx` from FW 6.61:
```
Debug: Reloc (null):1591 Type:X_HI16 Symbol:256 Offset 0000A848 Info:0000010D
Debug: Reloc (null):1592 Type:R_LO16 Symbol:256 Offset 0000A84C Info:00000106
Debug: Reloc (null):1593 Type:X_JAL26 Symbol:0 Offset 0000A888 Info:0000000F
Debug: Reloc (null):1594 Type:X_J26 Symbol:0 Offset 0000A8D0 Info:0000000E
```

**Comment by devnoname120** [comment]
If anyone wants to look into it, here is where relocations are done:
https://github.com/kotcrab/ghidra-allegrex/blob/e4717ee8bd624cdd1afbab4bf7434d7e9bacfcf1/src/main/kotlin/ghidra/app/util/bin/format/elf/relocation/AllegrexElfRelocationHandler.kt#L46-L73

**Comment by artart78** [comment]
The way to handle these relocations is specified in PRXTool.

Check https://github.com/pspdev/prxtool/blob/master/ProcessPrx.C#L881 to see how to parse them
and https://github.com/pspdev/prxtool/blob/master/ProcessPrx.C#L1439 to see how to do the actual relocation (I think the missing relocations are TypeB).

Handling these would be a great help for reversing kernel stuff!

**Comment by John-K** [comment]
@kotcrab I'm taking a stab at implementing this, but am having trouble logging things.
I'm able to build the plugin and copy / rename everything to get Ghidra to recognize it, but am unable to get any messages to print from the processor code.
I've tried added Msg.warn("blah") and println("blah") from AllegrexElfRelocationHandler.kt, but now output shows up in the "User Log" or otherwise - do you have any advice here?

**Comment by CelesteBlue-dev** [comment]
> @kotcrab I'm taking a stab at implementing this, but am having trouble logging things.
> I'm able to build the plugin and copy / rename everything to get Ghidra to recognize it, but am unable to get any messages to print from the processor code.
> I've tried added Msg.warn("blah") and println("blah") from AllegrexElfRelocationHandler.kt, but no output shows up in the "User Log" or otherwise - do you have any advice here?

I think you can use the println() function (maybe default JAVA library or Ghidra define) or the Msg library (but more for error messages).
I hope I understood well what you said.

**Comment by John-K** [comment]
If you re-read my message, I tried both of those and neither worked for me

**Comment by kotcrab** [comment]
`Msg.warn` etc. should show in user log but I'm not sure if I ever tested this, maybe some magic config file is missing.
`println()` will show in stdout. Not sure if it's required but I'm always running in debug mode for dev, launch through `./scripts/ghidraDebug`

If that fails you should just use the debug mode, it works well. Launch Ghidra in debug mode, open the project in IntelliJ IDEA (community version is perfectly fine), create `Remote JVM Debug` launch configuration and start it. Now you will be able to place breakpoints on any line and see variables etc. You can also use breakpoints for logging, right click it and you can uncheck suspend and enter log expression.

![image](https://user-images.githubusercontent.com/4594081/105392822-ee7d6600-5c1b-11eb-85db-e91accf19156.png)

> I'm able to build the plugin and copy / rename everything to get Ghidra to recognize it,

Instead of copying manually you can use `./gradlew ghidraInstall`, that will save you some time.

**Comment by John-K** [comment]
Ah, thanks for the pointer to debug mode. None of my messages print out still, but I do get a strange warning on stdout whenever I drag a new prx into the project:
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by net.sf.cglib.core.ReflectUtils$2 (file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Framework/Generic/lib/cglib-nodep-2.2.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of net.sf.cglib.core.ReflectUtils$2
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```
I guess I'll install IDEA and give that approach a shot.

**Comment by John-K** [comment]
I tried both Eclipse (couldn't quite get this to work) and IDEA can connect but none of my breakpoints get hit, despite the console showing that I'm connected and I'm able to pause ghidra.

IDEA Console shows:
`Connected to the target VM, address: 'localhost:18001', transport: 'socket'`

Ghidra thinks it has the Allegrex CPU loaded, and compiling in IDEA installs the plugin into the `Ghidra/Processor/Allegrex` folder

I'm not that familiar with Java / Kotlin development and feel like I'm fighting an uphill battle.

**Comment by John-K** [comment]
Ok they're being loaded at least (ran with `-verbose:class`)
```
[5.402s][info][class,load] ghidra.app.util.bin.format.elf.extend.Allegrex_ElfExtension source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.403s][info][class,load] ghidra.app.util.bin.format.MemoryLoadable source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Features/Base/lib/Base.jar
[5.403s][info][class,load] ghidra.app.util.bin.format.elf.ElfDynamicType source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Features/Base/lib/Base.jar
[5.404s][info][class,load] ghidra.app.util.bin.format.elf.ElfDynamicType$ElfDynamicValueType source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Features/Base/lib/Base.jar
[5.405s][info][class,load] ghidra.app.plugin.core.analysis.AllegrexAddressAnalyzer source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.406s][info][class,load] ghidra.app.plugin.core.analysis.AllegrexPreAnalyzer source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.406s][info][class,load] ghidra.app.util.opinion.PspElfLoader source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.407s][info][class,load] ghidra.app.util.opinion.PspElfLoader$Companion source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.408s][info][class,load] kotlin.Metadata source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.409s][info][class,load] kotlin.annotation.Retention source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.409s][info][class,load] kotlin.annotation.Target source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.410s][info][class,load] com.sun.proxy.$Proxy20 source: __dynamic_proxy__
[5.411s][info][class,load] ghidra.app.util.bin.format.elf.relocation.AllegrexElfRelocationHandler source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar
[5.411s][info][class,load] allegrex.format.elf.relocation.AllegrexRelocationFixupHandler source: file:/C:/Users/john/Desktop/ghidra_9.2.2_PUBLIC/Ghidra/Processors/Allegrex/lib/Allegrex.jar```

**Comment by kotcrab** [comment]
Your breakpoints look okay, line breakpoint icon has a tick to confirm it was correctly placed.

`Msg.warn` works for me so there seems to be some different issue if you can get any of those method to work.
![image](https://user-images.githubusercontent.com/4594081/105534708-2f45af80-5cee-11eb-9136-8e7376a2766f.png)

Are you sure the code you're trying to debug is actually executed? The relocations are processed after clicking OK in the import dialog. You can try adding a `error("")` (not `Msg.error`) or `throw IllegalStateException("")` to make it fail.

> I'm not that familiar with Java / Kotlin development and feel like I'm fighting an uphill battle.

Yeah. I appreciate you're trying to contribute to the project though.

> An illegal reflective access operation has occurred

You can ignore those.

> compiling in IDEA

I think you mean running Gradle task `ghidraInstall` which you can either run from command line or through IDEA. Doing compile in IDEA won't install it into Ghidra.

> Is there an easy way to know if Ghidra has loaded the Allegrex jar or if it's just loading the other files?

You can be extra sure if you delete your Allegrex dir and do `ghidraInstall` to make it reappear. Then dragging file should get detected as `PSP Executable` in import dialog.

**Comment by John-K** [comment]
Ok, so I'm able to add new messages to `parsePspRelocationTables()` in `PspElfHeader.kt` and `load()` in `PspElfLoader.kt and they print out:
```
2021-01-25 13:40:50 WARN  ghidra.app.util.opinion.PspElfLoader LOADED PSP BINARY  
2021-01-25 13:41:11 WARN  ghidra.app.plugin.core.analysis.AllegrexAddressAnalyzer ALLEGREX CAN ANALYZE!  
```
But nothing that I add in `AllegrexElfReloationExtension.kt`, ie `getType` gets printed out, nor in `canRelocate` in `AllegrexElfRelocationHandler.kt` - it seems like these classes aren't being used?

@kotcrab can you verify that added `Msg.warn` inside of those functions gets triggered for you from a clean checkout of `master`?

**Comment by John-K** [comment]
Ah, I just tried on a game PRX and things get triggered - something else is getting tripped up by Kernel PRXs.

I should have tested a game PRX earlier on... *sigh*

**Comment by kotcrab** [comment]
Oh I see, true, I only tested with games. 
Those classes are used but they won't have any references in code.

Will you try to figure this out by yourself? I don't really know where to find kernel PRXs, I guess I'd need to unpack some firmware but it's not something I'm familiar with.
At this point we need to check why Ghidra isn't using relocation handler from the plugin, there's is a bit of hacking going on there (see comment in `PspElfConstants.kt`). Though your file got detected by `PspElfLoader` so I can't really tell why without debugging.

**Comment by John-K** [comment]
I think I just figured out why:
`INFO  In parsePspRelocationTables, sections: 0, relocTables: 0  (PspElfHeader.kt:18)`

Which agrees with readelf:
```
readelf -Sr ~/PSP/Firmware/661/F0/kd/lowio.prx
There are no sections in this file.
There are no relocations in this file.
```

Will dig in further, I suspect it will require parsing some Sony-specific things from the LOAD header. 

Apologies for the spam. I'm just happy to be able to make progress and know that my tools are setup correctly.

**Comment by kotcrab** [comment]
I'm trying to implement this but I hit a wall. I need to get relocated file from prxtool to compare results of my code against prxtool. I compiled prxtool and I'm trying to relocate `lowio.prx` from FW 6.61 to `88000000`
```
./prxtool -r 2281701376 -e -o lowio.elf -d lowio.prx
...
Error: Invalid memory address for exports (0x0200B518)
Error: Couldn't load prx file structures
```
Okay... then I thought to add debug output of relocated instructions, I'd at least have some data to compare.
The first relocation is `R_MIPS_X_J26` so I added in `FixupRelocs`
```
COutput::Printf(LEVEL_DEBUG, "R_MIPS_X_J26 %08X %08X %08X\n", dwRealOfs, dwInst, dwData);
```
and got result
```
Debug: R_MIPS_X_J26 00000004 0800006E 0800886E
```
So 0800006E got turned into 0800886E, let's look in Ghidra:
First I loaded module without relocation applied:
![image](https://user-images.githubusercontent.com/4594081/132064130-3a55ab0c-8cdb-46f3-a26e-653f5281306b.png)
The jump makes sense.

Next manually relocated using prxtool value:
![image](https://user-images.githubusercontent.com/4594081/132064293-3f1db4ff-5393-4e3d-b5fa-3674c3fea2ff.png)
The jump is outside module bounds.

Looking at sources I don't understand how this can be valid for any jump instruction
https://github.com/pspdev/prxtool/blob/master/ProcessPrx.C#L1600

At this point I don't know if prxtool implementation is correct, maybe I'm doing something stupid. Ideally we would have dump of a relocated module from PSP memory to compare against but I have no idea how to do that on my own.

**Comment by kotcrab** [comment]
Blocked until someone provides working implementation of those relocations.

prxtool issue https://github.com/pspdev/prxtool/issues/6 (+I don't think their current relocations impl is correct)
uofw `sceKernelApplyPspRelSegment2` fails very early here https://github.com/uofw/uofw/blob/ef05dd20a3f43baa093452b4684346be4354543b/src/loadcore/loadelf.c#L808-L809 when feed `lowio.prx` data (I copied this to standalone program so maybe I did something wrong)

Fwiw you can take dumped memory and extract relocated modules then feed that into Ghidra.
![image](https://user-images.githubusercontent.com/4594081/132108983-5a0665db-bd5f-493f-bcd5-7eff1a083c42.png)
As you can see when loaded at `8808FD00` the first relocation `R_MIPS_X_J26` should result in `ae 3f 02 0a`

**Comment by kotcrab** [comment]
Update: we now have an implementation that's pretty likely correct. Though I'll need to work on it more before it can be merged into master.

**Comment by kotcrab** [comment]
Released v1.7 with support for this. Have fun reversing kernel modules, I regret I didn't get to this issue sooner.

Please create new issues if you spot any problems with this (feature ideas/requests are fine too btw). 
Unfortunately if there's any bug in relocation processing your existing project will likely be impossible to fix.
